### PR TITLE
Don't overwrite global err before return

### PR DESCRIPTION
**Pull request by oliverpool**
_Closes #846_ 

This PR adds a helper to overwrite the error returned by a function if it was previously nil and the deferred function returned a non-nil error.

I looked for occurrences of `defer func() {` and found a couple of packages where I this this helper improves the code.
Since there is no `util` package, I copied/paste the function : I am open to suggestion to reduce this code-duplication.

In `sqlutil`, I adapted the function to work for `EndTransaction`.

* [ ] I didn't add any new test (since this is not adding any feature), but feel free to hint any test that this PR could add
* [x] Pull request includes a [sign off](https://github.com/matrix-org/dendrite/blob/master/CONTRIBUTING.md#sign-off)

### CODE REVIEW

File: appservice/workers/transaction_scheduler.go [hunk 1]
```diff
@@ -207,7 +207,14 @@ func send(
 	appservice config.ApplicationService,
 	txnID int,
 	transaction []byte,
-) error {
+) (err error) {
+	defer func() {
+		if err != nil {
+			log.WithFields(log.Fields{
+				"appservice": appservice.ID,
+			}).WithError(err).Error("unable to send event")
```
Comment (Kegsay) [inline appservice/workers/transaction_scheduler.go:215]: I would rather log at the caller's side (the code which calls `send(...)`) than in a `defer` like this.

File: internal/sqlutil/sql.go [hunk 2]
```diff
@@ -38,9 +38,18 @@ type Transaction interface {
...
 	} else {
-		return txn.Rollback() // nolint: errcheck
+		return txn.Rollback()
+	}
+}
+
+// CheckEndTransactionErr ends a transaction and overwrites the error pointer if its value was nil.
+// If the transaction succeeded then it is committed, otherwise it is rolledback.
+// Designed to be used with defer (see EndTransaction otherwise).
+func CheckEndTransactionErr(txn Transaction, succeeded *bool, err *error) {
```
Comment (Kegsay) [inline internal/sqlutil/sql.go:50]: Slighty confusing naming: the `Check` implies that you've already called `EndTransaction` and are now checking the result, but we're actually calling `EndTransaction` here. `EndTransactionAndCheck` would be clearer but still awkward.

**Review by Kegsay — CHANGES_REQUESTED** [review]
Few minor things, otherwise LGTM thanks!

File: appservice/workers/transaction_scheduler.go (hunk 1 above)
Comment (oliverpool) [reply appservice/workers/transaction_scheduler.go]: :+1: done

File: internal/sqlutil/sql.go (hunk 2 above)
Comment (oliverpool) [reply internal/sqlutil/sql.go]: What do you think of `EndTransactionWithCheck`? (`With` is quite usual in the standard library)

Comment (Kegsay) [reply internal/sqlutil/sql.go]: Sounds good to me.

**Review by Kegsay — APPROVED** [review]
Next time please do not force push changes to PRs, as it makes it impossible for me to see the diff from when I last reviewed.

**Review by neilalexander — APPROVED** [review]

**Pull request merged**